Fix #6180 (Usage of variable after std::move or std::forward)

zingsheim opened the pull request:
I want to put this pull request under discussion.
It is a first version and there are still open questions.

The proposed version finds access of moved or forwarded variables.

Some examples can be seen in test/testmove.cpp.

Furthermore I want to discuss the outcome at some real world code, i.e. facebook/folly.
In folly the developers have a lot of test whether the result of a moved variable has the expected results after the move. This is OK for tests but in ordinary code a user should not rely on the outcome of a moved variable.
``` c++
TEST(small_vector, NonCopyableType) {
  folly::small_vector<NontrivialType,2> vec;

  for (int i = 0; i < 10; ++i) {
    vec.emplace(vec.begin(), 13);
  }
  EXPECT_EQ(vec.size(), 10);
  auto vec2 = std::move(vec);
  EXPECT_EQ(vec.size(), 0);				// <- Access of moved variable vec.
  EXPECT_EQ(vec2.size(), 10);
  vec2.clear();

  folly::small_vector<NoncopyableCounter,3> vec3;
  for (int i = 0; i < 10; ++i) {
    EXPECT_EQ(vec3.size(), i);
    EXPECT_EQ(NoncopyableCounter::alive, i);
    vec3.insert(vec3.begin(), NoncopyableCounter());
  }
  EXPECT_EQ(vec3.size(), 10);
  EXPECT_EQ(NoncopyableCounter::alive, 10);

  vec3.insert(vec3.begin() + 3, NoncopyableCounter());
  EXPECT_EQ(NoncopyableCounter::alive, 11);
  auto vec4 = std::move(vec3);
  EXPECT_EQ(NoncopyableCounter::alive, 11);
  vec4.resize(30);
  EXPECT_EQ(NoncopyableCounter::alive, 30);
  vec4.erase(vec4.begin(), vec4.end());
  EXPECT_EQ(vec4.size(), 0);
  EXPECT_EQ(NoncopyableCounter::alive, 0);
}
```

Another example is the following. Also here one can see a test of the variable after it has been moved. Additionally, there is a the assign of the vector. The proposed commit does not know that by the assign function the vector is totally overwritten independent of its content.
``` c++
TEST(fbvector, move_construction) {
  fbvector<int> v1(100, 100);
  fbvector<int> v2;
  EXPECT_EQ(v1.size(), 100);
  EXPECT_EQ(v1.front(), 100);
  EXPECT_EQ(v2.size(), 0);
  v2 = std::move(v1);
  EXPECT_EQ(v1.size(), 0);			// <- Access of moved variable v1.
  EXPECT_EQ(v2.size(), 100);
  EXPECT_EQ(v2.front(), 100);

  v1.assign(100, 100);				// <- Access of moved variable v1.
  auto other = std::move(v1);			// <- Access of moved variable v1.
  EXPECT_EQ(v1.size(), 0);
  EXPECT_EQ(other.size(), 100);
  EXPECT_EQ(other.front(), 100);
}
```
There are a lot of error messages to code like this in folly.

Now follows two really strange examples (which work in folly only due to the special usage of the functions).

``` c++
template<class C>
void testRangeFunc(C&& x, size_t n) {
  const auto& cx = x;
  // type, conversion checks
  Range<int*> r1 = range(std::forward<C>(x));
  Range<const int*> r2 = range(std::forward<C>(x));   // <- Access of moved variable x.
  Range<const int*> r3 = range(cx);                   // <- Access of moved variable x.
  Range<const int*> r5 = range(std::move(cx));        // <- Access of moved variable x.
  EXPECT_EQ(r1.begin(), dataPtr(x));
  EXPECT_EQ(r1.end(), dataPtr(x) + n);
  EXPECT_EQ(n, r1.size());
  EXPECT_EQ(n, r2.size());
  EXPECT_EQ(n, r3.size());
  EXPECT_EQ(n, r5.size());
}
```
``` c++
template<class F, class Tuple>
struct GuardObj : GuardObjBase {
  explicit GuardObj(F&& f, Tuple&& args)
    : f_(std::forward<F>(f))
    , args_(std::forward<Tuple>(args))
  {}
  GuardObj(GuardObj&& g) noexcept
    : GuardObjBase(std::move(g))
    , f_(std::move(g.f_))				// <- Access of moved variable g.
    , args_(std::move(g.args_))				// <- Access of moved variable g.
  {}

  ~GuardObj() {
    folly::applyTuple(f_, args_);
  }

  GuardObj(const GuardObj&) = delete;
  GuardObj& operator=(const GuardObj&) = delete;

private:
  F f_;
  Tuple args_;
};
```

Finally, I want to show an example, where the check cannot know that ```tryPutMessageNoThrow``` only take over the moved variable when it returns ```true```.
``` c++
void AsyncServerSocket::dispatchError(const char *msgstr, int errnoValue) {
  uint32_t startingIndex = callbackIndex_;
  CallbackInfo *info = nextCallback();

  // Create a message to send over the notification queue
  QueueMessage msg;
  msg.type = MessageType::MSG_ERROR;
  msg.err = errnoValue;
  msg.msg = std::move(msgstr);

  while (true) {
    // Short circuit if the callback is in the primary EventBase thread
    if (info->eventBase == nullptr) {
      std::runtime_error ex(
        std::string(msgstr) +  folly::to<std::string>(errnoValue));
      info->callback->acceptError(ex);
      return;
    }

    if (info->consumer->getQueue()->tryPutMessageNoThrow(std::move(msg))) {
      return;
    }
    // Fall through and try another callback

    if (callbackIndex_ == startingIndex) {
      // The notification queues for all of the callbacks were full.
      // We can't really do anything at this point.
      LOG(ERROR) << "failed to dispatch accept error: all accept callback "
        "queues are full: error msg:  " <<
        msg.msg.c_str() << errnoValue;	                                    // <- Access of moved variable msg. 
      return;
    }
    info = nextCallback();
  }
}
```

Now, to the questions I want to discuss:
1. Should the error message distinguish between moved and forwarded variables?
2. Should the error message distinguish between moved and possibly moved variables, i.e. variables which were moved only in an if branch?
3. Which functions should be detected as moves (and forward, respectively)?
 a. Any function named move
 b. Only std::move
 c. std::move or move but not name::move or name.move (I do not know if it is possible to see the active namespaces)
4. I have added the detection of the move to ```valueFlowAfterAssign``` since the behavior is very similar. I treat a move as an assignment with a for the user unknown value, since the user cannot rely on the result after the move. Is this OK?
5. I have added the move check to a completely new source, i.e. checkmove.cpp. Unfortunately, the check requires an iteration through the whole source. Should the check be included into another already existing iteration though the source?
6. For which settings the check should be activated?
7. How should the message be classified error/style/... ?

PKEuS commented:
This check is so small, so that I think that it should be moved to CheckOther

danmar commented:
Thanks!

My knowledge about exactly how std::move and std::forward works is a bit limited.

1. I don't know. I don't think that would hurt.
2. that is not necessary. we say "division by zero" if you assign zero to a variable in an if and then later divide with that variable.
3. I would say b. I doubt "move" is a unique function name.
4. sounds ok. didn't look at the code yet though.
5. it should do an independant iteration through the source.
6,7. This is about UB, isn't it? Then the severity is either error or warning. If you know that the object is taken in a path so there is UB then you can write "error". If you have some tryPutMessageNoThrow(std::move(msg)) then as far as I know, it just looks suspicious and a warning is more appropriate. however we will run daca2 and if we see that such function calls are often safe in real code then I don't want a warning at all.

danmar commented on lib/checkmove.cpp:
```diff
@@ -0,0 +1,63 @@
+void CheckMove::checkAccess()
+{
+    for (const Token *tok = _tokenizer->tokens(); tok; tok = tok->next()) {
+        if (tok->values.empty())
```
this if is useless. can it be faster to avoid the loop below?

danmar commented on lib/checkmove.cpp:
```diff
@@ -0,0 +1,63 @@
+    for (const Token *tok = _tokenizer->tokens(); tok; tok = tok->next()) {
+        if (tok->values.empty())
+            continue;
+        for (std::list<ValueFlow::Value>::const_iterator it=tok->values.begin(); it!=tok->values.end(); ++it) {
```
I somehow think this should be in Token since the other value lookup functions are there.

zingsheim replied:
Can you point out which part of Token you are referring to? I did not find which part of Token handling value lookup you are talking about.

danmar replied:
The token has : hasKnownIntValue() , getValue() , getMaxValue() , getValueLE(), getValueGE() , getValueTokenMaxStrLength() , getValueTokenMinStrSize() , getValueTokenDeadPointer() .. add one more method here.

danmar commented on lib/checkmove.cpp:
```diff
@@ -0,0 +1,63 @@
+
+void CheckMove::AccessMovedError(const Token *tok, const std::string &varname, bool inconclusive)
+{
+    const std::string errmsg(std::string("Access of moved variable ") + varname + ".");
```
I suggest to shorten this

    const std::string errmsg("Access of moved variable " + varname + ".");

danmar commented on lib/checkmove.h:
```diff
@@ -0,0 +1,91 @@
+    }
+private:
+
+    void AccessMovedError(const Token *tok, const std::string &varname, bool inconclusive);
```
nit: please use lowercase for first letter. accessMovedError()

danmar commented on lib/valueflow.cpp:
```diff
@@ -1551,27 +1589,6 @@ static bool valueFlowForward(Token * const               startToken,
                 }
             }
 
-            // bailout if address of var is taken..
```
as far as I see this code was moved. why?

zingsheim replied:
The reason is that the value flow does not set a value into the out parameter of a function.
```
void g(A & a);
void f() {
    ...
    g(b);
    ...
}
```
In the example above it prevents that the value flow sets the value of ```b``` in the function call even if the value of ```b``` is known. The reason for this is that ```b``` might be overwritten by ```g``` anyhow. I hope other parts of the code do not make use of the value to make further checks inside ```g```.
For the value flow of the move this is needed. Otherwise the following code gives a wrong warning that the moved variable a is accessed in the line of ```g(a)``` in the following code:
```
void g(A & a);
void f() {
    A a;
    A b = std::move(a);
    g(a);
    A c = a;
}
```
Note: This example if also part of the new test ```moveAndFunctionParameter2()```

If for some reason the original order is needed then it must be checked in ```CheckMove::checkAccess()``` if the moved variable is an (in-)out parameter of a function. This is possible, too.

zingsheim replied:
I wrote:
> If for some reason the original order is needed then it must be checked in CheckMove::checkAccess() if the moved variable is an (in-)out parameter of a function. This is possible, too.

The function valueFlowSubFunction does make use of the value of the argument. Even though it currently does not do it for non-const reference parameters since value flow bails out in this case in 
```
static void valueFlowInjectParameter(TokenList* tokenlist, ErrorLogger* errorLogger, const Settings* settings, const Variable* arg, const Scope* functionScope, const std::list<ValueFlow::Value>& argvalues)
{
    // Is argument passed by value or const reference, and is it a known non-class type?
    if (arg->isReference() && !arg->isConst() && !arg->isClass())
        return;
```

But I do not want to block an improvement here.
Therefore, I will keep the original order of the code in lib/valueflow.cpp and instead add a check in ```CheckMove::checkAccess()``` if the moved variable is an (in-)out parameter of a function.

PS: Maybe PKEuS can give some insights why valueFlowInjectParameter bails out for non-const reference parameters.

zingsheim replied:
If I remove the bailout in ```valueFlowInjectParameter``` due to the non-const reference parameter no test object fails.
However, I can construct an example were the behavior without the bailout gives better results, i.e.
```
void dostuff(int & x) {
  f(x);
}
void test() {        
    static int x;
    if (a) x = 2;
    dostuff(x); 
}
```
Without the bailout I get the following debug output:
```
##file test.cpp
1: void dostuff ( int & x@1 ) {
2: f ( x@1 ) ;
3: }
4: void test ( ) {
5: static int x@2 ;
6: if ( a ) { x@2 = 2 ; }
7: dostuff ( x@2 ) ;
8: }



##Value flow
Line 2
  x possible 2
Line 6
  2 always 2
Line 7
  x possible 2
```
With the bailout the value flow information is reduced, i.e. the possible x=2 value on line 2 is missing.
```
##Value flow
Line 6
  2 always 2
Line 7
  x possible 2
```

danmar replied:
thanks for that info. interesting.

danmar commented on lib/valueflow.cpp:
```diff
@@ -1596,52 +1613,107 @@ static bool valueFlowForward(Token * const               startToken,
+    return nullptr;
+}
+
+static const Token * findEndOfStatement(Token * tok)
```
this will crash if tok->astOperand2() is null somehow

danmar commented on lib/valueflow.cpp:
```diff
@@ -1596,52 +1613,107 @@ static bool valueFlowForward(Token * const               startToken,
+                // Assignment
+
+                // Lhs should be a variable
+                if (!tok->astOperand1() || !tok->astOperand1()->varId()) {
```
It's not a hard rule but I don't use {} around a single expression like this.

danmar commented on lib/valueflow.cpp:
```diff
@@ -1596,52 +1613,107 @@ static bool valueFlowForward(Token * const               startToken,
-                else break;
-            }
-            nextExpression = nextExpression->next();
+                std::vector<FlowBarierType>::reverse_iterator foundFlowBarier = find_if(flowBariers.rbegin()+1, flowBariers.rend(), FlowBarierVaridPredicate(varid));
```
I prefer full name std::find_if .. this is not portable there is not a "using namespace std;" anywhere.

danmar commented:
After a little thinking I would prefer if you create a new valueflow function .. for instance valueFlowMove(). So valueFlowAfterAssign() specifically handles assignments.

Then you will also need to add more guard logic so when isMoved is set then intvalue/tokvalue should not be used. I want to add more types of values later .. so I think we should consider adding some enum like:

    enum ValueType { Int, Tok, Moved, Float, Iterator, ... };

danmar commented:
> After a little thinking I would prefer if you create a new valueflow function .. for instance valueFlowMove(). So valueFlowAfterAssign() specifically handles assignments.

I change my mind again. You should probably keep this in valueFlowAfterAssign(). I need to look at it closer and understand it better.

I think it looks promising.

> Finally, I want to show an example, where the check cannot know that tryPutMessageNoThrow only take over the moved variable when it returns true.

Thanks.. according to the Cppcheck philosophy we will not warn about that. A warning might be ok if it's highly likely that there is a bug.

For information I have been thinking about using whole-program-analysis in ValueFlow. The new --cppcheck-build-dir is a possible first step.

zingsheim commented:
I have worked in the comments from above and made a few other changes.

1. The check is only applied for >=cpp11

2. The inconclusive handling is added

3. I check the implementation on https://github.com/apple/swift. Due to this I have done following improvements/corrections:

- Value flow of assignment in return statement
- Do not handle ```a``` as moved in statements like ```std::move(a).getX()```, since it only move out parts of the structure and further move out of other parts like ```std::move(a).getY()```are still possible. 

The check on swift points out two suspicious code segments in swift due to move:

- In "swift/lib/SILGen/SILGenApply.cpp"
Note: In the marked line below the parameter ```callee``` which is already been moved is used and not the member variable ```callee```.
```
    CallEmission(SILGenFunction &gen, Callee &&callee,
                 WritebackScope &&writebackScope,
                 bool assumedPlusZeroSelf = false)
      : gen(gen),
        callee(std::move(callee)),
        InitialWritebackScope(std::move(writebackScope)),
        uncurries(callee.getNaturalUncurryLevel() + 1),           // <<-- Access of moved variable callee
        applied(false),
        AssumedPlusZeroSelf(assumedPlusZeroSelf)
    {
```

- In "swift/lib/SILGen/SILGenLValue.cpp"
```
    ManagedValue existentialAddr
      = gen.emitAddressOfLValue(e, std::move(existentialLV), accessKind);

    // Open up the existential.
    LValue lv;
    lv.add<ValueComponent>(existentialAddr, existentialLV.getTypeData());     // <<-- Access of moved variable existentialAddr
```

If you know another open source project with extensive use of the move semantics, I can check this to, in order to harden the pull request implementation. 

PS: Do you know why recent build crash in the TestThreadExecutor::many_threads on windows?

zingsheim commented:
I have updated the pull request once more.
I have made more cases inconclusive to have not a false positive in the following code which might be quite common:
```
std::vector<int> v;
v.push(1);    // push some data
...
f(std::move(v));  // f might swap an internal vector with v
v.clear();        // clear the vector and reuse allocated memory inside
v.push(100);      // push some new data
...
```

danmar commented:
Thanks for working on this.

I changed the ValueFlow to allow other types of values except intvalue/tokvalue. 473266748847daa663021902a561a0f02f6ffe64.

Please adapt your code. Add an enum constant MOVED so your values can't be used as intvalues by mistake.

zingsheim commented:
I have rebased the patch on the top of the latest commit and worked in the new ValueFlow::Value::MOVED.

On top of this I have created another commit. "Improve std::move in lambda capture list by AST of capture list". I have added this as a separate commit such that you can have a look at it whether the addition of the AST inside the lambda capture list is OK. The handling of the lambda is by far not complete especially due to the varId setting.
I have now added this since I saw a usage of std::move inside the capture list in "https://github.com/chriskohlhoff/asio".

danmar commented on lib/checkother.h:
```diff
@@ -252,6 +256,8 @@ class CPPCHECKLIB CheckOther : public Check {
+    void accessMovedError(const Token *tok, const std::string &varname, ValueFlow::Value::MoveKind moveKind, bool inconclusive);
 
     void getErrorMessages(ErrorLogger *errorLogger, const Settings *settings) const {
         CheckOther c(nullptr, settings, errorLogger);
```
also update the classInfo()

zingsheim replied:
Done.

danmar commented on lib/valueflow.cpp:
```diff
@@ -1692,52 +1725,127 @@ static bool valueFlowForward(Token * const               startToken,
+                start = memberInitializationTok;
+        }
+
+        std::vector<FlowBarrierType> flowBarriers;
```
I am not sure I like these changes.. need to look carefully and understand it better.

danmar replied:
ok I understand. I agree there should be unit tests for this in TestValueFlow. I believe the TestOther should just test that if a variable has a MOVED value and is used wrongly then a warning is written.

danmar replied:
somehow when I look at it .. it does not look natural that std::move() is handled in valueFlowAfterAssign().

Maybe we could add a valueFlowAfterMove() such as:

    void valueFlowAfterMove() {
        if (!isAtLeastCpp11)
            return;
        for (Token* tok = start; tok != scope->classEnd; tok = tok->next()) {
            if (!isStdMoveOrStdForwarded(tok, &moveKind, &varTok))
                continue;
            // x is not MOVED after assignment if code is:  x = ... std::move(x) .. ;
            const Token *parent = tok->astParent();
            while (parent && parent->str() != "=")
                 parent = parent->astParent();
            if (parent && parent->astOperand2()->str() == varTok->str())
                 continue;
             valueFlowForward...
        }
    }

I am hoping this will mean that we can keep valueFlowAfterAssign as it is.

zingsheim commented on lib/tokenlist.cpp:
```diff
@@ -976,6 +977,23 @@ static void compileExpression(Token *&tok, AST_state& state)
+        return false;
+    return true;
+}
+
```
Before, I had a different logic for the test inside isLambdaCaptureList but I realized that it was not completely correct.
The check now is consistent with the definition of the AST.
However, the AST might change in future to include ```x``` and ```y``` in the AST.
Furthermore a lambda without parameter ```y``` (i.e. ```[x](){}```) can also be written shorter as ```[x]{}``` (see http://en.cppreference.com/w/cpp/language/lambda (4)). Is this handled correctly?

danmar replied:
hmm.. yes the ast might change.. that should make it more correct.

zingsheim commented:
I worked in your suggestions, except the one for rightmostLeaf (see comment above).

zingsheim commented:
I changed the implementation to your suggestion of ```valueFlowAfterMove```.
The code of ```valueFlowAfterAssign``` look now cleaner again.
With this separation future extensions of the one or the other function are probably much easier to reason about.

I also added some early returns in ```isStdMoveOrStdForwarded``` even though I am not a fan of early return statements since they might have a surprising effect when you add some logic later and you wounder why your code is never executed in same situations.

danmar commented on lib/checkother.cpp:
```diff
@@ -2697,3 +2697,78 @@ void CheckOther::unknownEvaluationOrder(const Token* tok)
+{
+    if (!_tokenizer->isCPP() || _settings->standards.cpp < Standards::CPP11)
+        return;
+    bool reportInconclusive = _settings->inconclusive;
```
minor nit.. it would be better with const.

danmar commented on lib/valueflow.cpp:
```diff
@@ -1692,16 +1705,115 @@ static bool valueFlowForward(Token * const               startToken,
+        const Scope * scope = symboldatabase->functionScopes[i];
+        const Token * start = scope->classStart;
+        if (scope->function) {
+            const Token * memberInitializationTok = scope->function->constructorMemberInitialization();
```
as far as I see this is redundant. The valueFlowAfterAssign() will not handle the initializer list as far as I see.

danmar commented on lib/tokenlist.cpp:
```diff
@@ -976,6 +977,23 @@ static void compileExpression(Token *&tok, AST_state& state)
         compileComma(tok, state);
 }
 
+static bool isLambdaCaptureList(Token * tok)
```
please use const

danmar commented on lib/tokenlist.cpp:
```diff
@@ -1060,25 +1078,30 @@ static Token * createAstAtToken(Token *tok, bool cpp)
 
         // Compile inner expressions inside inner ({..}) and lambda bodies
         for (tok = tok1->next(); tok && tok != endToken; tok = tok ? tok->next() : nullptr) {
-            if (tok->str() != "{")
```
please don't remove this early continue.

danmar replied:
hmm.. ok now I see why you removed it.

danmar commented on lib/valueflow.cpp:
```diff
@@ -1811,9 +1811,15 @@ static void valueFlowAfterAssign(TokenList *tokenlist, SymbolDatabase* symboldat
                 start = memberInitializationTok;
         }
 
+        const Token * isInForCondition = nullptr;   // if in for condition then pointer to end of for condition
```
hmm  .. this isInForCondition is not used as far as I see.

danmar commented:
Thanks! I applied your changes with minor tweaks with c8ff96fe8fe85a71d32705bde170bfa81ba6926e.
